File: backend/src/main/java/cn/etarch/mao/harness/core/PersistedChatMessage.java
package cn.etarch.mao.harness.core;

import cn.etarch.mao.harness.llm.ChatRequest;

/** Internal message representation that keeps persistence state out of the LLM protocol DTO. */
public record PersistedChatMessage(
        Long messageId,
        String persistedContentSnapshot,
        ChatRequest.Message chatMessage) {

    public PersistedChatMessage(Long messageId, ChatRequest.Message chatMessage) {
        this(messageId, TokenEstimator.contentToString(chatMessage.getContent()), chatMessage);
    }
}

File: backend-ts/src/harness/core/persisted-chat-message.ts
import type { ChatMessage } from '../llm/chat-request.js';
import { TokenEstimator } from './token-estimator.js';

export class PersistedChatMessage {
  constructor(
    readonly messageId: number,
    readonly persistedContentSnapshot: string,
    readonly chatMessage: ChatMessage,
  ) {}

  static from(messageId: number, chatMessage: ChatMessage, snapshot?: string): PersistedChatMessage {
    return new PersistedChatMessage(
      messageId,
      snapshot ?? TokenEstimator.contentToString(chatMessage.content),
      chatMessage,
    );
  }
}
